### PR TITLE
Forking processes for xbutil2

### PR DESCRIPTION
**Pull request by AShivangi**
**This completes the functionality implementation of xbutil2 validate**

- Created a standalone Process.h to abstract the implementation of "runPythonScript" from SubCmdValidate
- Added new nodes "Xclbin" and "Testcase" in the log for debugging purposes
- Cleanup (removed unwanted warning suppressions), restructured code and documentation 

Tests:
- 1RP  on ubuntu 16.04
- 2RP on ubuntu 16.04
- ubuntu18.04 (boost::process) 
- Not tested on windows since it requires the shell directory to be defined but the code compiles 

Sample output:
```
$ xbutil --new validate -d 0000:b3:00.1
-----------------------------------------------------
 Invoking next generation of the xbutil application
-----------------------------------------------------
Starting validation for 1 devices

Validate device[0000:b3:00.1]
Platform            : xilinx_u200_xdma_201830_3
SC Version          : 4.3.9
Platform ID         : 0x5ee83d83

1/10 Test #1         : Kernel version
    Description     : Check if kernel version is supported by XRT
    [PASSED]
-------------------------------------------------------------------------------
2/10 Test #2         : Aux connection
    Description     : Check if auxiliary power is connected
    Warning         : Aux power is not connected
    Warning         : Device is not stable for heavy acceleration tasks
    [PASSED WITH WARNINGS]
-------------------------------------------------------------------------------
3/10 Test #3         : PCIE link
    Description     : Check if PCIE link is active
    [PASSED]
-------------------------------------------------------------------------------
4/10 Test #4         : SC version
    Description     : Check if SC firmware is up-to-date
    [PASSED]
-------------------------------------------------------------------------------
5/10 Test #5         : Verify kernel
    Description     : Run 'Hello World' kernel test
    Xclbin          : /opt/xilinx/dsa/xilinx_u200_xdma_201830_3/test/verify.xclbin
    Testcase        : /opt/xilinx/xrt/test/22_verify.py
    [PASSED]
-------------------------------------------------------------------------------
6/10 Test #6         : DMA
    Description     : Run dma test
    Details         : Host -> PCIe -> FPGA write bandwidth = 9064.513845 MB/s
    Details         : Host <- PCIe <- FPGA read bandwidth = 11737.398630 MB/s
    [PASSED]
-------------------------------------------------------------------------------
7/10 Test #7         : Bandwidth kernel
    Description     : Run 'bandwidth kernel' and check the throughput
    Xclbin          : /opt/xilinx/dsa/xilinx_u200_xdma_201830_3/test/bandwidth.xclbin
    Testcase        : /opt/xilinx/xrt/test/23_bandwidth.py
    Details         : Maximum throughput: 48073 MB/s
    [PASSED]
-------------------------------------------------------------------------------
8/10 Test #8         : Peer to peer bar
    Description     : Run P2P test
    Details         : P2P bar is not enabled
    [SKIPPED]
-------------------------------------------------------------------------------
9/10 Test #9         : Memory to memory DMA
    Description     : Run M2M test
    Details         : M2M is not available
    [SKIPPED]
-------------------------------------------------------------------------------
10/10 Test #10        : Host memory bandwidth test
    Description     : Run 'bandwidth kernel' when slave bridge is enabled
    Details         : Address translator IP is not available
    [SKIPPED]
-------------------------------------------------------------------------------
Validation completed, but with warnings
```

### CODE REVIEW

File: src/runtime_src/core/tools/common/Process.cpp [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+  is_done = true;
+  bool exitCode = (os_stdout.str().find("PASS") != std::string::npos) ? true : false;
+  run_test.get()->finish(exitCode, "");
+  // Workaround: Clear the default progress bar output so as to print the Error: before printing [FAILED]
+  // Remove this once busybar is implemented
+  std::cout << EscapeCodes::cursor().prev_line() << EscapeCodes::cursor().clear_line();
+  t.join();
+
+  return exitCode ? 0 : 1;
```
Comment (keryell) [inline src/runtime_src/core/tools/common/Process.cpp:160]: `!exitCode`

File: src/runtime_src/core/tools/common/Process.cpp [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+  bool is_done = false;
+  // Fix: create busy bar
+  auto run_test = std::make_shared<XBUtilities::ProgressBar>("Running Test", max_test_duration, XBUtilities::is_esc_enabled(), std::cout); 
+  std::thread t(testCaseProgressReporter, run_test, std::ref(is_done));
+
+  // Close existing stderr and set it to be the write end of the pipe.
+  // After fork below, our child process's stderr will point to the same fd.
+  dup2(stderr_fds[1], STDERR_FILENO);
+  close(stderr_fds[1]);
+  std::shared_ptr<FILE> stderr_child(fdopen(stderr_fds[0], "r"), fclose);
```
Comment (keryell) [inline src/runtime_src/core/tools/common/Process.cpp:128]: How does the destructor work?

File: src/runtime_src/core/tools/common/Process.cpp [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+  int counter = 0;
+  while(counter < max_test_duration && !is_done) {
+    run_test.get()->update(counter);
+    counter++;
+    std::this_thread::sleep_for(std::chrono::seconds(1));
+  }
+}
+
+unsigned int
+XBUtilities::runPythonScript( const std::string & script,
```
Comment (keryell) [inline src/runtime_src/core/tools/common/Process.cpp:89]: Random formatting

File: src/runtime_src/core/tools/common/Process.cpp [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+  // Find the python executable
+  boost::filesystem::path pythonAbsPath = boost::process::search_path("py");  
+  if (pythonAbsPath.string().empty()) 
+    pythonAbsPath = boost::process::search_path("python");   
+
+  if (pythonAbsPath.string().empty()) 
+    throw std::runtime_error("Error: Python executable not found in search path.");
+
+  // Make sure the script exists
+  if ( !boost::filesystem::exists( script ) ) {
```
Comment (keryell) [inline src/runtime_src/core/tools/common/Process.cpp:179]: Random formatting.

File: src/runtime_src/core/tools/common/Process.cpp [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+  boost::process::environment env = boost::this_process::environment();
+  env.erase("XCL_EMULATION_MODE");
+
+  // Please fix: Should be a busy bar and NOT a progress bar
+  ProgressBar run_test("Running Test", max_test_duration, XBUtilities::is_esc_enabled(), std::cout); 
+
+  // Execute the python script and capture the outputs
+  boost::process::ipstream ip_stdout;
+  boost::process::ipstream ip_stderr;
+  boost::process::child runningProcess( pythonAbsPath,
```
Comment (keryell) [inline src/runtime_src/core/tools/common/Process.cpp:202]: Random formatting

File: src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp [hunk 6]
```diff
@@ -343,63 +189,84 @@ searchLegacyXclbin(const std::string& dev_name, const std::string& xclbin, boost
...
-    if(xclbinPath.empty()) {
-      if(xclbin.compare("bandwidth.xclbin") == 0) {
-        //if bandwidth xclbin isn't present, skip the test
-        logger(_ptTest, "Details", "Bandwidth xclbin not available. Skipping validation.");
-        _ptTest.put("status", "skipped");
-      }
-      return;
+  //check if a 2RP platform
+  std::vector<std::string> logic_uuid;
+  try{
```
Comment (keryell) [inline src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:214]: Random formatting

File: src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp [hunk 7]
```diff
@@ -343,63 +189,84 @@ searchLegacyXclbin(const std::string& dev_name, const std::string& xclbin, boost
...
+    logic_uuid = xrt_core::device_query<xrt_core::query::logic_uuids>(_dev);
+  } catch(...) { }
+
+  std::string xclbinPath;
+  if(!logic_uuid.empty()) {
```
Comment (keryell) [inline src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:219]: Random formatting

File: src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp [hunk 8]
```diff
@@ -343,63 +189,84 @@ searchLegacyXclbin(const std::string& dev_name, const std::string& xclbin, boost
...
-    std::string xrtTestCasePath = "/opt/xilinx/xrt/test/" + py;
-    boost::filesystem::path xrt_path(xrtTestCasePath);
-    if (!boost::filesystem::exists(xrt_path)) {
-      logger(_ptTest, "Error", boost::str(boost::format("Failed to find %s") % xrtTestCasePath));
-      logger(_ptTest, "Error", "Please check if the platform package is installed correctly");
-      _ptTest.put("status", "failed");
-      return;
+  //check if xclbin is present
+  if(xclbinPath.empty()) {
+    if(xclbin.compare("bandwidth.xclbin") == 0) {
```
Comment (keryell) [inline src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:227]: Random formatting

File: src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp [hunk 9]
```diff
@@ -343,63 +189,84 @@ searchLegacyXclbin(const std::string& dev_name, const std::string& xclbin, boost
...
+    logger(_ptTest, "Error", os_stdout.str());
+    logger(_ptTest, "Error", os_stderr.str());
+    _ptTest.put("status", "failed");
+  } 
+  else {
+    _ptTest.put("status", "passed");
+  }
+
+  // Get out max thruput for bandwidth testcase
+  if(xclbin.compare("bandwidth.xclbin") == 0) {
```
Comment (keryell) [inline src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:263]: Random formatting

**Comment by stsoe** [comment]
@AShivangi Please include me in review of makefile changes.

**Pull request merged**